daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/FNFunctions.scala

```diff
@@ -803,6 +805,84 @@ case class FNFloor(recipe: CompiledDPath, argType: NodeInfo.Kind) extends FNOneA
   }
 }
 
+case class DFDLXLeftShift(recipes: List[CompiledDPath], argType: NodeInfo.Kind) extends FNTwoArgs(recipes) {
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+
+    val shiftLong = arg2.getLong
+    val shift = shiftLong.toInt
+    Assert.invariant(shift >= 0)
+    argType match {
+      case NodeInfo.Long => {
+
+        if (shift >= 64) dstate.SDE("Left shift not supported for shift value greater than 64") else DataValue.toDataValue(arg1.getLong << shift)
+      }
+      case NodeInfo.Int => {
+        if (shift >= 32) dstate.SDE("Left shift not supported for shift value greater than 32") else DataValue.toDataValue(arg1.getInt << shift)
+      }
+      case NodeInfo.Short => {
+        if (shift >= 16) dstate.SDE("Left shift not supported for shift value greater than 16") else DataValue.toDataValue((arg1.getShort << shift).toShort)
+      }
+      case NodeInfo.Byte => {
+        if (shift >= 8) dstate.SDE("Left shift not supported for shift value greater than 8") else DataValue.toDataValue((arg1.getByte << shift).toByte)
+      }
+      case NodeInfo.UnsignedLong =>{
+        if(shift >=64) dstate.SDE("Left shift not supported for shift value greater than 64") else DataValue.toDataValue((arg1.getBigInt << shift).toLong)
```

Comment:
Unsigned long is represented by a java BigInteger, not a Long, so I think toLong here is incorrect. If the argument is an UnsignedLong (Java BigInteger) then we need to return a Java BigInteger. Same issue with below unsigned types. Need to make sure the underlying type stays the same. 

Some additional concerns:
1. I thought the shift operator doesn't work for BigInteger. Does this need to be ``arg1.getBigInt.shiftLeft(shift)``? I would expect a compliation error here, but maybe there's some scala implicit magic going on?
2. BigInteger has no bounds, so a shift left could go beyond the bounds of an UnsignedLong. A similar issue exists for the other unsigned cases as well. For example, for UnsignedInt, we get the Long then shift which could create a value larger than can fit in an UnsignedInt. I think we need to mask off any bits the should normally be shifted off. Or as Mike suggested, maybe we should convert these values to the unsigned classes we have (e.g. ULong), do the shift, and then convert back to correct primitive (e.g. BigInt).

This is some very subtly complicated code.

Reply:
@stevedlawrence I got your points.
@mbeckerle Due to some emergency, I need to travel somewhere.
I will continue to work on this from Monday, and will try to finish it as soon as possible.
Thank You.